HARK/ConsumptionSaving/ConsWealthPortfolioModel.py

```diff
@@ -1,624 +1,438 @@
+    top_c = end_dvda_nvrs[a_idx, share_idx + 1]
+    bot_dvda = end_dvda[a_idx, share_idx]
+    top_dvda = end_dvda[a_idx, share_idx + 1]
+    alpha = 1.0 - top_f / (top_f - bot_f)
+
+    # Calculate the continuous optimal risky share and optimal consumption
+    Share_now = (1.0 - alpha) * bot_s + alpha * top_s
+    end_dvda_nvrs_now = (1.0 - alpha) * bot_c + alpha * top_c
+    end_dvda_now = (1.0 - alpha) * bot_dvda + alpha * top_dvda
+
+    # If agent wants to put more than 100% into risky asset, he is constrained.
+    # Likewise if he wants to put less than 0% into risky asset, he is constrained.
+    constrained_top = focs[:, -1] > 0.0
+    constrained_bot = focs[:, 0] < 0.0
+
+    # Apply those constraints to both risky share and consumption (but lower
+    # constraint should never be relevant)
+    Share_now[constrained_top] = 1.0
+    Share_now[constrained_bot] = 0.0
+    end_dvda_nvrs_now[constrained_top] = end_dvda_nvrs[constrained_top, -1]
+    end_dvda_nvrs_now[constrained_bot] = end_dvda_nvrs[constrained_bot, 0]
+    end_dvda_now[constrained_top] = end_dvda[constrained_top, -1]
+    end_dvda_now[constrained_bot] = end_dvda[constrained_bot, 0]
+
+    # When the natural borrowing constraint is *not* zero, then aNrm=0 is in the
+    # grid, but there's no way to "optimize" the portfolio if a=0, and consumption
+    # can't depend on the risky share if it doesn't meaningfully exist. Apply
+    # a small fix to the bottom gridpoint (aNrm=0) when this happens.
+    if not BoroCnstNat_iszero:
+        Share_now[0] = 1.0
+        end_dvda_nvrs_now[0] = end_dvda_nvrs[0, -1]
+        end_dvda_now[0] = end_dvda[0, -1]
+
+    # Construct functions characterizing the solution for this period
+
+    # Now this is where we look for optimal C
+    # for each a in the agrid find corresponding c that satisfies the euler equation
+
+    cNrm_now = np.empty_like(aNrmGrid)
+
+    for a_idx, a_nrm in enumerate(aNrmGrid):
+        cNrm_guess = euler(
+            aNrmGrid,
+            a_nrm,
+            CRRA,
+            WealthShare,
+            WealthShift,
+            end_dvda_now[a_idx],
         )
 
-        self.solution.shareEndFunc = self.shareEndFunc
+        lower_bound = aNrmGrid[cNrm_guess > 0][-1]
+        upper_bound = aNrmGrid[cNrm_guess < 0][0]
 
-    def solve(self):
-        # Make arrays of end-of-period assets and end-of-period marginal values
-        self.prepare_to_calc_EndOfPrd()
-        if self.method in ["root", "fxp"]:
-            self.calc_EndOfPrdvP()
-        elif self.method == "max":
-            self.calc_EndOfPrdv()
+        result = root_scalar(
+            euler,
+            method="toms748",
+            bracket=[lower_bound, upper_bound],
+            args=(a_nrm, CRRA, WealthShare, WealthShift, end_dvda_now[a_idx]),
+        )
 
-        # Construct a basic solution for this period
-        self.prepare_to_optimize_share()
-        self.optimize_risky_share()
+        cNrm_now[a_idx] = result.root
```

Comment:
@mnwhite could you give me your thoughts on this re-worked solver? 

This problem does not have an invertible euler equation, so I instead try to do root-finding

everything else should be the same as the simple portfolio problem, and I have confirmed that when WealthShare=0.0, it is indeed the same result

Reply:
might be easier to just see the solver rather than try to see the diffs 

https://github.com/econ-ark/HARK/blob/51e1d357617406d0b482fcb8f2613c9d2bd94af2/HARK/ConsumptionSaving/ConsWealthPortfolioModel.py